### PR TITLE
TFLite and Pytorch CPP launcher, fps measure change and minor updates

### PR DESCRIPTION
**Pull request by maslovaz**
- change FPS metrics calcuation: fps = overall num of images inferred / overall time of inference for all python launchers
- add batch fps metrics: batch_fps = batch size / median_latency
- add json report to all python launchers
- add pytorch c++ launcher
- add tensorflow Lite c++ launcher
- move latency calculation to dedicated decorator

### CODE REVIEW

File: src/benchmark/README.md [hunk 1]
```diff
@@ -90,9 +90,11 @@ Inference Engine предоставляет 2 программных интер
 - **Латентность (latency)** - медиана множества времен выполнения вывода.
 - **Среднее время одного прохода (average time of single pass)** - отношение
   общего времени выполнения всех запросов к числу запросов.
-- **Количество кадров, обрабатываемых за секунду (frames per second, FPS)** -
+- **Количество кадров, обрабатываемых за секунду для пачки (batch frames per second, BATCH FPS)** -
   отношение размера обрабатываемой "пачки" изображений к среднему времени
   одного прохода.
+- **Среднее количество кадров, обрабатываемое за секунду(frames per second, FPS)** -
```
Comment (valentina-kustikova) [inline src/benchmark/README.md:96]: Если я правильно поняла, то договорились, что теперь этот показатель будет использоваться для обоих режимов (и для синхронного, и для асинхронного), но введен он только в блоке, где описываются показатели для синхроннного интерфейса. Что не так?

File: src/benchmark/README.md [hunk 2]
```diff
@@ -101,11 +103,12 @@ Inference Engine предоставляет 2 программных интер
...
   итераций цикла тестирования. Характеризует время выполнения набора
   одновременно созданных запросов на устройстве.
-- **Количество кадров, обрабатываемых за секунду (frames per second, FPS)** -
+- **Среднее количество кадров, обрабатываемых за секунду (frames per second, FPS)** -
   отношение произведения размера обрабатываемой "пачки" и числа итераций
   к общему времени выполнения всех запросов.
 
-**Примечание:** в публикуемой html-таблице содержатся только показатели FPS.
+**Примечание:** в публикуемой html-таблице содержатся только показатели FPS (до 06.2023 в таблице содержались
+показатели Batch FPS).
```
Comment (valentina-kustikova) [inline src/benchmark/README.md:111]: Надо еще добавить комментарии на Wiki после того, как замержим PR.

File: src/configs/README.md [hunk 3]
```diff
@@ -92,6 +92,10 @@
     по каждому из каналов входного изображения.
   - `InputScale`- тег, необязательный для заполнения; только для случая, когда `frontend` не равен `ir`;
     может отсуствовать. Определяет коэффициент масштабирования входного изображения.
+  - `ChangePreprocessOptions`- тег, необязательный для заполнения; только для случая, когда `frontend` не равен `ir`
+    и используется версия OpenVINO >= 2023.0.0; может отсуствовать. 
+    Command line может отличаться для разных версий benchmark_app (Дает возможность переименования
```
Comment (valentina-kustikova) [inline src/configs/README.md:97]: Что имеется ввиду под Command line?

После скобки заглавная буква

File: src/configs/README.md [hunk 4]
```diff
@@ -157,6 +161,11 @@
     физическому количеству ядер в системе.
   - `InferenceRequestsCount` - тег, необязательный для заполнения. Определяет число запросов на вывод. По умолчанию
     не установлен и выбирается фреймворком автоматически.
+  - `InputType` - тег, необязательный для заполнения; используется в PyTorch CPP. Определяет формат входных данных
+  (в виде `[FP32]` или `input1[F32],input2[FP32]`)
```
Comment (valentina-kustikova) [inline src/configs/README.md:165]: Форматирование, отступ пропал

File: src/cpp_dl_benchmark/pytorch_launcher/README.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
+# Pytorch Benchmark
+
+The tool allows to measure deep learning models inference performance with C++ distribution of [PyTorch][pytorch].
+
+## Build Pytorch
+
+You can download C++ distribution of PyTorch from official [site][pytorch] (cxx11 ABI needed, for Nvidia GPU select distribution for CUDA), or build it from sources ([official instruction][build-instruction]):
```
Comment (valentina-kustikova) [inline src/cpp_dl_benchmark/pytorch_launcher/README.md:7]: Не больше 80 символов в строку.

NVIDIA

File: src/cpp_dl_benchmark/pytorch_launcher/README.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+1. Clone Pytorch repo:
+    ```
```
Comment (valentina-kustikova) [inline src/cpp_dl_benchmark/pytorch_launcher/README.md:9]: bash после трех кавычек

File: src/cpp_dl_benchmark/pytorch_launcher/README.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
+# Pytorch Benchmark
+
+The tool allows to measure deep learning models inference performance with C++ distribution of [PyTorch][pytorch].
+
+## Build Pytorch
+
+You can download C++ distribution of PyTorch from official [site][pytorch] (cxx11 ABI needed, for Nvidia GPU select distribution for CUDA), or build it from sources ([official instruction][build-instruction]):
+1. Clone Pytorch repo:
```
Comment (valentina-kustikova) [inline src/cpp_dl_benchmark/pytorch_launcher/README.md:8]: Лучше добавлять пустую строку после пункта, а то где-то так, а где-то по-другому написано

File: src/cpp_dl_benchmark/tflite_launcher/README.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+## Build TensorFlow Lite
+
+Clone repository, checkout to the `2.10.0` release:
+```
+git clone  https://github.com/tensorflow/tensorflow
+cd tensorflow
+git checkout v2.10.0
+```
+
+### CMake build
```
Comment (valentina-kustikova) [inline src/cpp_dl_benchmark/tflite_launcher/README.md:14]: Пустая строка после заголовка

File: src/cpp_dl_benchmark/tflite_launcher/README.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    >libabsl_time.so
+    >libabsl_time_zone.so
+    >```
+    >
+
+### Bazel build
+
+Second opiton is to build TF Lite with bazel.
+
+1. (Optional) Install python3 and essential modules:
```
Comment (valentina-kustikova) [inline src/cpp_dl_benchmark/tflite_launcher/README.md:102]: Пустая строка после пункта

File: src/cpp_dl_benchmark/tflite_launcher/README.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    sudo apt update && sudo apt full-upgrade
+    bazel --version
+
+    ```
+
+1. Go to `tensorflow` directory:
+
+    ```
+    cd tensorflow
+    ```
```
Comment (valentina-kustikova) [inline src/cpp_dl_benchmark/tflite_launcher/README.md:127]: Пустая строка после кода

**Comment by valentina-kustikova** [comment]
@FenixFly, @a-sidorova, просьба, оперативно посмотреть текущий PR.

File: README.md [hunk 11]
```diff
@@ -74,11 +74,11 @@ Novgorod State University Publishing House, 2021. – 423 p.
...
 - `results` directory contains benchmarking and validation results.
 
-  - [`benchmarking`](results/benchmarking) contains benchmarking 
+  - [`benchmarking`](results/benchmarking) contains benchmarking
     results in html- and xslx-formats.
   - [`accuracy`](results/accuracy) contains accuracy
     results in html- and xslx-formats.
-  - [`validation`](results/validation) contains tables that confirms 
+  - [`validation`](results/validation) contains tables that confirms
     correctness of inference implementation for the benchmarked models.
```
Comment (valentina-kustikova) [inline README.md:82]: При описании директории `cpp_dl_benchmark` надо вписать полный перечень доступных фреймворков (C++ API).

**Comment by valentina-kustikova** [comment]
@maslovaz, @n-berezina-nn, просьба, после мержа текущего PR обновить страницы на Wiki. Спасибо!

File: src/configs/README.md [hunk 12]
```diff
@@ -157,6 +161,11 @@
...
+  - `TensorRTPrecision` - тег, необязательный для заполнения; используется в PyTorch CPP. Определяет тип данных весов
+    для случая машины с NVIDIA GPU и PyTorch, построенного с поддержкой TensorRT.
+    Возможные значения: FP16, FP32. По умолчанию не установлен.
 
 - Набор тегов для тестирования вывода средствами TensorFlow Lite:
```
Comment (FenixFly) [inline src/configs/README.md:170]: TensorFlow Lite и TensorFlow Lite CPP?

File: src/configs/benchmark_configuration_file_template.xml [hunk 13]
```diff
@@ -236,17 +237,51 @@
...
+            <Layout></Layout>
             <Mean></Mean>
             <Std></Std>
             <OutputName></OutputName>
             <ModelType></ModelType>
             <InferenceMode></InferenceMode>
+            <TensorRTPrecision></TensorRTPrecision>
+        </FrameworkDependent>
+    </Test>
+    <Test>
```
Comment (FenixFly) [inline src/configs/benchmark_configuration_file_template.xml:256]: Кажется также не хватает примеров конфига для TFLite и TFLite CPP

**Review by FenixFly — CHANGES_REQUESTED** [review]

File: src/configs/README.md (hunk 12 above)
Comment (maslovaz) [reply src/configs/README.md:170]: конкретно тут - питоновый АПИ, обновлю

File: test/smoke_test/smoke_config.xml [hunk 14]
```diff
@@ -17,8 +17,8 @@
             <InferenceFramework>OpenVINO DLDT</InferenceFramework>
```
[the remark this answers is not in the document]
Comment (maslovaz) [reply test/smoke_test/smoke_config.xml:17]: мы измеряем через benchmark_app, поддержка как плюсового, так и питонового benchmark_app из опенвино уже давно есть

File: src/cpp_dl_benchmark/common/utils/src/utils.cpp [hunk 15]
```diff
@@ -26,14 +27,22 @@ utils::Device utils::get_device_from_str(const std::string& dstr) {
...
 
-std::string utils::get_precision_str(DataPrecision p) {
+std::string get_data_precision_str(const DataPrecision p) {
     if (precision_to_str_map.count(p) > 0) {
         return precision_to_str_map.at(p);
     }
     return "UNKNOWN";
 }
+DataPrecision get_data_precision_from_str(const std::string& pstr) {
+    for (const auto [data_precision, precision_str] : precision_to_str_map) {
```
[the remark this answers is not in the document]
Comment (maslovaz) [reply src/cpp_dl_benchmark/common/utils/src/utils.cpp]: спасибо, добавлено

File: src/cpp_dl_benchmark/pytorch_launcher/src/pytorch_launcher.cpp [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    }
+    else {
+        std::ostringstream s;
+        s << type;
+        throw std::invalid_argument("Does not support element data type " + s.str());
+    }
+}
+
+torch::Dtype get_data_type(utils::DataPrecision precision) {
+    for (const auto [type, data_prectision] : dtype_to_precision_map) {
```
[the remark this answers is not in the document]
Comment (maslovaz) [reply src/cpp_dl_benchmark/pytorch_launcher/src/pytorch_launcher.cpp]: спасибо, добавлено

File: src/inference/README.md [hunk 17]
```diff
@@ -10,6 +10,7 @@
 1. TensorFlow Lite.
 1. MXNet.
 1. PyTorch.
+1. ONNX Runtime.
```
[the remark this answers is not in the document]
Comment (maslovaz) [reply src/inference/README.md:13]: done

File: src/configs/benchmark_configuration_file_template.xml (hunk 13 above)
Comment (maslovaz) [reply src/configs/benchmark_configuration_file_template.xml:256]: added

File: src/configs/README.md (hunk 12 above)
Comment (maslovaz) [reply src/configs/README.md]: done

File: README.md (hunk 11 above)
Comment (maslovaz) [reply README.md:82]: done

File: src/cpp_dl_benchmark/tflite_launcher/README.md (hunk 10 above)
Comment (maslovaz) [reply src/cpp_dl_benchmark/tflite_launcher/README.md:131]: done

File: src/cpp_dl_benchmark/tflite_launcher/README.md (hunk 8 above)
Comment (maslovaz) [reply src/cpp_dl_benchmark/tflite_launcher/README.md:15]: done

File: src/cpp_dl_benchmark/pytorch_launcher/README.md (hunk 7 above)
Comment (maslovaz) [reply src/cpp_dl_benchmark/pytorch_launcher/README.md:11]: done

File: src/cpp_dl_benchmark/pytorch_launcher/README.md (hunk 6 above)
Comment (maslovaz) [reply src/cpp_dl_benchmark/pytorch_launcher/README.md]: done

File: src/configs/README.md (hunk 4 above)
Comment (maslovaz) [reply src/configs/README.md]: done

File: src/cpp_dl_benchmark/pytorch_launcher/README.md (hunk 5 above)
Comment (maslovaz) [reply src/cpp_dl_benchmark/pytorch_launcher/README.md]: done

File: src/configs/README.md (hunk 3 above)
Comment (maslovaz) [reply src/configs/README.md]: переписала немного текст

File: src/benchmark/README.md (hunk 1 above)
Comment (maslovaz) [reply src/benchmark/README.md]: проблема только в ридми, обновила

File: src/benchmark/README.md [hunk 18]
```diff
@@ -90,9 +90,11 @@ Inference Engine предоставляет 2 программных интер
 - **Латентность (latency)** - медиана множества времен выполнения вывода.
 - **Среднее время одного прохода (average time of single pass)** - отношение
   общего времени выполнения всех запросов к числу запросов.
-- **Количество кадров, обрабатываемых за секунду (frames per second, FPS)** -
+- **Количество кадров, обрабатываемых за секунду для пачки (batch frames per second, BATCH FPS)** -
   отношение размера обрабатываемой "пачки" изображений к среднему времени
   одного прохода.
+- **Среднее количество кадров, обрабатываемое за секунду(frames per second, FPS)** -
+  отношение полного количества обработанных изображений к полному времени инференса
```
[the remark this answers is not in the document]
Comment (maslovaz) [reply src/benchmark/README.md]: done

File: src/benchmark/frameworks/intel_caffe/intel_caffe_process.py [hunk 19]
```diff
@@ -33,7 +28,9 @@ def _fill_command_line(self):
         device = self._test.indep_parameters.device
         iteration = self._test.indep_parameters.iteration
 
-        common_params = f'-m {model_prototxt} -w {model_caffemodel} -i {dataset} -b {batch} -d {device} -ni {iteration}'
+        common_params = (f'-m {model_prototxt} -w {model_caffemodel} -i {dataset} -b {batch} -d {device}'
```
[the remark this answers is not in the document]
Comment (n-berezina-nn) [reply src/benchmark/frameworks/intel_caffe/intel_caffe_process.py:31]: с точки зрения ланчера это опциональный аргумент, но с точки зрения запуска inference_benchmark скрипта это параметр обязательный, тк из файла отчета выпарсиваются значения метрик. и без передачи этого параметра это не получится, либо где-то хардкодить и поддерживать консистентность дефолтных названий репортов в ланчере и обертках, что тоже плохо.

[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/benchmark/frameworks/intel_caffe/intel_caffe_process.py:31]: У меня возник данный комментарий, потому что в PR'e в большинстве скриптов написано через `self._add_optional_argument_to_cmd_line`, а в этом в одном месте сразу эксплиситно. Мы можем тогда, пожалуйста, выравнить поведение?

Спасибо!

File: test/smoke_test/smoke_config.xml (hunk 14 above)
[the remark this answers is not in the document]
Comment (a-sidorova) [reply test/smoke_test/smoke_config.xml:17]: Я, может, неправильно задала вопрос или ваш ответ не понимаю, но я имела в виду, что у нас smoke tests проверяют лишь python inference для фреймворков в то время, когда у нас есть поддержка C++ для того же PyTorch и TFLite теперь. Хотелось бы также проверять работоспособность и данных лаунчеров. Что думаете?

[the remark this answers is not in the document]
Comment (maslovaz) [reply test/smoke_test/smoke_config.xml:17]: Я думала, что вопрос именно про опенвино=)
обсудили вчера с @valentina-kustikova - тестирование плюсовых ланчеров в рамках github actions-плохая затея. у нас есть такой же смок тест во внутреннем CI на плюсовых ланчерах

File: src/benchmark/frameworks/intel_caffe/intel_caffe_process.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (maslovaz) [reply src/benchmark/frameworks/intel_caffe/intel_caffe_process.py:31]: @a-sidorova
выровняла

**Review by a-sidorova — APPROVED** [review]

**Comment by maslovaz** [comment]
@FenixFly всё еще висит статус "requested changes"

**Comment by maslovaz** [comment]
@valentina-kustikova посмотрите, пожалуйста, финально еще разок

**Review by FenixFly — APPROVED** [review]

File: src/benchmark/README.md [hunk 20]
```diff
@@ -101,11 +103,12 @@ Inference Engine предоставляет 2 программных интер
   отношение времени выполнения всех наборов запросов к числу
   итераций цикла тестирования. Характеризует время выполнения набора
   одновременно созданных запросов на устройстве.
-- **Количество кадров, обрабатываемых за секунду (frames per second, FPS)** -
+- **Среднее количество кадров, обрабатываемых за секунду (frames per second, FPS)** -
```
Comment (valentina-kustikova) [inline src/benchmark/README.md:106]: По смыслу (если я правильно поняла) FPS для обоих режимов считается одинаково, а определения даются разные. Поправьте меня, пожалуйста. Если я права, то имеет смысл написать одинаковые определения для FPS и называть их по-русски одинаково.

**Review by valentina-kustikova — COMMENTED** [review]
@maslovaz, все ок. Написала один вопрос.

File: src/benchmark/README.md (hunk 20 above)
Comment (valentina-kustikova) [reply src/benchmark/README.md:106]: Мне Github показывает следующее:

Для latency-режима:
- **Среднее количество кадров, обрабатываемое за секунду (frames per second, FPS)** -
  отношение полного количества обработанных изображений к полному времени вывода.

Для troughput-режима:
- **Среднее количество кадров, обрабатываемых за секунду (frames per second, FPS)** -
  отношение произведения размера обрабатываемой "пачки" и числа итераций
  к общему времени выполнения всех запросов.

Comment (maslovaz) [reply src/benchmark/README.md]: Нашла, не то сравнивала.
Спасибо, исправила!

**Review by valentina-kustikova — APPROVED** [review]

**Comment by maslovaz** [comment]
@valentina-kustikova сегодня-завтра обновлю вики

**Comment by valentina-kustikova** [comment]
> @valentina-kustikova сегодня-завтра обновлю вики

@maslovaz, спасибо!
